### PR TITLE
MON-3211: Implement switching to metrics-server

### PR DESCRIPTION
**Pull request by slashpai**
As part of epic https://issues.redhat.com/browse/MON-3153

**Note:** Not all options of Prometheus Adapter are implemented in Metrics Server in this PR

**(I will add ChangeLog once everything finalized to avoid rebase issues)**

* [ ] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

**Comment by slashpai** [comment]
Most of the failures is due to image references which can done after updating https://github.com/openshift/kubernetes-metrics-server

File: pkg/manifests/manifests.go [hunk 1]
```diff
@@ -2079,6 +2091,75 @@ func (f *Factory) PrometheusAdapterAPIService() (*apiregistrationv1.APIService,
...
+
+func (f *Factory) MetricsServerClusterRoleBindingAuthDelegator() (*rbacv1.ClusterRoleBinding, error) {
+	return f.NewClusterRoleBinding(f.assets.MustNewAssetReader(MetricsServerClusterRoleBindingAuthDelegator))
+}
+
+func (f *Factory) MetricsServerRoleBindingAuthReader() (*rbacv1.RoleBinding, error) {
+	return f.NewRoleBinding(f.assets.MustNewAssetReader(MetricsServerRoleBindingAuthReader))
+}
+
+func (f *Factory) MetricsServerDeployment() (*appsv1.Deployment, error) {
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:2118]: The TLS settings should be injected at runtime based on the API global settings like we do here for prometheus-adapter
https://github.com/openshift/cluster-monitoring-operator/blob/ebaa7e3bc9fd26ed8566a046d29a61d4e9db6c3e/pkg/manifests/manifests.go#L2013-L2014

File: jsonnet/components/metrics-server.libsonnet [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        },
+        spec: {
+          affinity: {
+            podAntiAffinity: {
+              requiredDuringSchedulingIgnoredDuringExecution: [
+                {
+                  labelSelector: {
+                    matchLabels: {
+                      'app.kubernetes.io/name': 'metrics-server',
+                      'app.kubernetes.io/component': 'metrics-server',
```
Comment (simonpasquier) [inline jsonnet/components/metrics-server.libsonnet:195]: (nit) we should try to be consistent between affinity and pod (L171) selectors.

File: jsonnet/components/metrics-server.libsonnet [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+        },
+      },
+      template: {
+        metadata: {
+          labels: {
+            'app.kubernetes.io/name': 'metrics-server',
+            'app.kubernetes.io/component': 'metrics-server',
+          } + cfg.commonLabels,
+        },
+        spec: {
```
Comment (simonpasquier) [inline jsonnet/components/metrics-server.libsonnet:187]: I think that we discussed and agreed on using minReadySeconds to ensure that metrics availability on upgrades?

File: jsonnet/components/metrics-server.libsonnet [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+                  port: 'https',
+                  scheme: 'HTTPS',
+                },
+                initialDelaySeconds: 20,
+                periodSeconds: 10,
+              },
+              resources: {
+                requests: {
+                  cpu: '100m',
+                  memory: '200Mi',
```
Comment (simonpasquier) [inline jsonnet/components/metrics-server.libsonnet:250]: We'd need to adjust these numbers following https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#resources-and-limits

File: jsonnet/components/metrics-server.libsonnet [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        'app.kubernetes.io/component': 'metrics-server',
+      } + cfg.commonLabels,
+      name: 'metrics-server',
+      namespace: cfg.namespace,
+    },
+    spec: {
+      minAvailable: 1,
+      selector: {
+        matchLabels: {
+          'app.kubernetes.io/name': 'metrics-server',
```
Comment (simonpasquier) [inline jsonnet/components/metrics-server.libsonnet:318]: same here regarding selector consistency

**Comment by slashpai** [comment]
still working on changes moved out of draft to test

**Comment by slashpai** [comment]
Need to wait till https://github.com/openshift/cluster-monitoring-operator/pull/2151 is merged and rebased

**Comment by slashpai** [comment]
I will rebase once https://github.com/openshift/cluster-monitoring-operator/pull/2151  is merged for now add commits to check this whole implementation

File: pkg/manifests/manifests.go [hunk 6]
```diff
@@ -2098,6 +2112,89 @@ func (f *Factory) PrometheusAdapterAPIService() (*apiregistrationv1.APIService,
...
+	containers[idx].Args = f.setTLSSecurityConfiguration(podSpec.Containers[0].Args,
+		MetricsServerTLSCipherSuitesFlag, MetricsServerTLSMinTLSVersionFlag)
+
+	config := f.config.ClusterMonitoringConfiguration.MetricsServerConfig
+	if config == nil {
+		return dep, nil
+	}
+
+	if len(config.NodeSelector) > 0 {
+		podSpec.NodeSelector = config.NodeSelector
```
Comment (slashpai) [inline pkg/manifests/manifests.go:2164]: q: For all the config fields if not specified do we want to copy the value from prometheus-adapter deployment?

File: assets/metrics-server/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+metadata:
+  labels:
+    app.kubernetes.io/component: metrics-server
+    app.kubernetes.io/managed-by: cluster-monitoring-operator
+    app.kubernetes.io/name: metrics-server
+    app.kubernetes.io/part-of: openshift-monitoring
+  name: metrics-server
+  namespace: openshift-monitoring
+spec:
+  minReadySeconds: 60
```
Comment (slashpai) [inline assets/metrics-server/deployment.yaml:12]: q: In enhancement proposal we have mentioned "Wait for `5 x metrics-server's scrape interval` to have enough data populated in `metrics-server` in-memory`" to switch APIService

` --metric-resolution` set is `15s`,  do we want to make this 75 seconds instead of 60 seconds?

File: hack/build-jsonnet.sh [hunk 8]
```diff
@@ -37,7 +37,7 @@ for file in "${files[@]}"; do
 	}&
 
 	# wait for at least one of the jobs to finish if there are more than maxProc jobs
-	while [[ $(jobs -r | wc -l ) -ge "$maxProc" ]]; do wait -n; done
+	while [[ $(jobs -r | wc -l ) -ge "$maxProc" ]]; do wait; done
```
Comment (simonpasquier) [inline hack/build-jsonnet.sh:40]: why this change?

File: jsonnet/components/metrics-server.libsonnet [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    metadata: {
+      labels: {
+        'app.kubernetes.io/name': 'metrics-server',
+        'app.kubernetes.io/component': 'metrics-server',
+      } + cfg.commonLabels,
+      name: 'metrics-server',
+      namespace: cfg.namespace,
+    },
+  },
+  clusterRoleAggregatedMetricsReader: {
```
Comment (simonpasquier) [inline jsonnet/components/metrics-server.libsonnet:16]: can we move it to the cluster-monitoring-operator component to avoid duplication with prometheus-adapter?

File: jsonnet/components/metrics-server.libsonnet [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+      minAvailable: 1,
+      selector: {
+        matchLabels: {
+          'app.kubernetes.io/name': 'metrics-server',
+          'app.kubernetes.io/component': 'metrics-server',
+        } + cfg.commonLabels,
+      },
+    },
+  },
+  apiService: {
```
Comment (simonpasquier) [inline jsonnet/components/metrics-server.libsonnet:326]: same remark here: shouldn't it be moved to cluster-monitoring-operator?

File: jsonnet/components/metrics-server.libsonnet [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+              },
+              name: 'metrics-server',
+              ports: [
+                {
+                  containerPort: 10250,
+                  name: 'https',
+                  protocol: 'TCP',
+                },
+              ],
+              readinessProbe: {
```
Comment (simonpasquier) [inline jsonnet/components/metrics-server.libsonnet:239]: It would be good to explain the probe settings (e.g. how it correlates to scrape intervals).

File: pkg/manifests/config.go [hunk 12]
```diff
@@ -207,6 +208,10 @@ func NewConfig(content io.Reader, tp bool) (*Config, error) {
 		return nil, errors.Wrap(ErrConfigValidation, "collectionProfiles is a TechPreview feature, to be able to use a profile different from the default (\"full\") please enable TechPreview")
 	}
 
+	if c.ClusterMonitoringConfiguration.MetricsServerConfig != nil && !tp {
```
Comment (simonpasquier) [inline pkg/manifests/config.go:211]: we need to rely on the MetricsServer feature gate.

File: pkg/manifests/manifests.go (hunk 6 above)
Comment (simonpasquier) [reply pkg/manifests/manifests.go:2164]: eventually yes but it can/should be a follow-up PR.

File: assets/metrics-server/deployment.yaml (hunk 7 above)
Comment (simonpasquier) [reply assets/metrics-server/deployment.yaml:12]: From https://github.com/kubernetes-sigs/metrics-server/pull/768, the readiness probe should return OK only when there are enough samples in memory so it's fine to remove `minReadySeconds` IMHO.

**Comment by slashpai** [comment]
> Looking at the techpreview job, I see strange errors in the metrics-server logs (https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-monitoring-operator/2022/pull-ci-openshift-cluster-monitoring-operator-master-e2e-aws-ovn-techpreview/1727222590193274880/artifacts/e2e-aws-ovn-techpreview/gather-extra/artifacts/pods/openshift-monitoring_metrics-server-64597c87fd-xnv7f_metrics-server.log):
> 
> ```
> E1122 07:57:54.881962       1 scraper.go:140] "Failed to scrape node" err="Get \"https://10.0.14.96:10250/metrics/resource\": context deadline exceeded" node="ip-10-0-14-96.us-west-2.compute.internal"
> E1122 07:58:09.885476       1 scraper.go:140] "Failed to scrape node" err="Get \"https://10.0.93.247:10250/metrics/resource\": context deadline exceeded" node="ip-10-0-93-247.us-west-2.compute.internal"
> E1122 07:58:24.881469       1 scraper.go:140] "Failed to scrape node" err="Get \"https://10.0.93.247:10250/metrics/resource\": context deadline exceeded" node="ip-10-0-93-247.us-west-2.compute.internal"
> E1122 07:58:39.881828       1 scraper.go:140] "Failed to scrape node" err="Get \"https://10.0.93.247:10250/metrics/resource\": context deadline exceeded" node="ip-10-0-93-247.us-west-2.compute.internal"
> E1122 07:58:54.888048       1 scraper.go:140] "Failed to scrape node" err="Get \"https://10.0.93.247:10250/metrics/resource\": context deadline exceeded" node="ip-10-0-93-247.us-west-2.compute.internal"
> ```

Looking at kubelet error metric around same time (checked metrics from  prometheus using promecieus) kubelet did have issues around that time. When I check kubelet logs on TechPreview cluster kubelet is seen restarted couple of times even without this change

kubelet_runtime_operations_errors_total

File: jsonnet/components/metrics-server.libsonnet (hunk 10 above)
Comment (slashpai) [reply jsonnet/components/metrics-server.libsonnet:326]: Since this depends on service we cannot move this one atleast for now

File: hack/build-jsonnet.sh (hunk 8 above)
Comment (slashpai) [reply hack/build-jsonnet.sh:40]: ok `-n` is invalid option for mac terminal so I needed to remove that while running generate command, accidentally went in change :)

File: jsonnet/components/metrics-server.libsonnet (hunk 9 above)
Comment (slashpai) [reply jsonnet/components/metrics-server.libsonnet:16]: I created separate PR https://github.com/openshift/cluster-monitoring-operator/pull/2163 since that includes more changes too

File: pkg/tasks/prometheusadapter.go [hunk 13]
```diff
@@ -11,24 +11,34 @@ import (
 )
 
 type PrometheusAdapterTask struct {
-	client    *client.Client
-	ctx       context.Context
-	factory   *manifests.Factory
-	config    *manifests.Config
-	namespace string
+	client               *client.Client
+	metricsServerEnabled bool
```
Comment (simonpasquier) [inline pkg/tasks/prometheusadapter.go:15]: for consistency with the metrics-server task, I would have the flag defining that prometheus-adapter is enabled
```suggestion
	enabled bool
```

File: hack/build-jsonnet.sh (hunk 8 above)
Comment (simonpasquier) [reply hack/build-jsonnet.sh:40]: could you revert then?

File: pkg/manifests/config.go [hunk 14]
```diff
@@ -207,6 +208,10 @@ func NewConfig(content io.Reader, tp bool) (*Config, error) {
 		return nil, errors.Wrap(ErrConfigValidation, "collectionProfiles is a TechPreview feature, to be able to use a profile different from the default (\"full\") please enable TechPreview")
 	}
 
+	if c.ClusterMonitoringConfiguration.MetricsServerConfig != nil && !metricsServerEnabled {
```
Comment (simonpasquier) [inline pkg/manifests/config.go:211]: since we have a dedicated feature gate for metrics server, I wouldn't error out if the config is present. It can even be preferred to handle upgrades (e.g. you can setup the metrics server config before it goes GA).

File: pkg/tasks/prometheusadapter.go (hunk 13 above)
Comment (jan--f) [reply pkg/tasks/prometheusadapter.go:15]: Using this flag the p-a resource removal could also be handled here instead of the metricsserver task.

File: pkg/tasks/metricsserver.go [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+		}
+	}
+	{
+		err := pa.client.DeleteDeployment(ctx, d)
+		if err != nil {
+			return errors.Wrap(err, "deleting PrometheusAdapter Deployment failed")
+		}
+	}
+
+	// TODO(slashpai): Add steps to remove other resources
```
Comment (jan--f) [inline pkg/tasks/metricsserver.go:208]: Is this `TODO` still valid? I was wondering about the p-a `APIService`. I thought there can only be on registered `APIService` for a given `group`. I might be missing details there though.

Comment (slashpai) [reply pkg/tasks/metricsserver.go:208]: I put that as TODO to do recheck :). I think we will have some more roles to remove

**Comment by slashpai** [comment]
I will rebase once https://github.com/openshift/cluster-monitoring-operator/pull/2163 is merged. I thought that was merged as it was ready to merge state :(. CI seems to have issues

File: pkg/tasks/metricsserver.go (hunk 15 above)
Comment (slashpai) [reply pkg/tasks/metricsserver.go]: API Service we are updating in metrics server task create step. Cannot delete that one since same name

**Comment by slashpai** [comment]
cc: @juzhao

**Comment by slashpai** [comment]
This is good for review again, addressed comments

File: pkg/tasks/prometheusadapter.go (hunk 13 above)
Comment (slashpai) [reply pkg/tasks/prometheusadapter.go]: we would still need to call the delete method from metrics server since we need to wait till metrics-server is ready to serve requests

File: pkg/tasks/prometheusadapter.go [hunk 16]
```diff
@@ -12,15 +12,17 @@ import (
...
 	client    *client.Client
+	enabled   bool
 	ctx       context.Context
 	factory   *manifests.Factory
 	config    *manifests.Config
 	namespace string
 }
 
-func NewPrometheusAdapterTask(ctx context.Context, namespace string, client *client.Client, factory *manifests.Factory, config *manifests.Config) *PrometheusAdapterTask {
+func NewPrometheusAdapterTask(ctx context.Context, namespace string, client *client.Client, prometheusAdapterEnabled bool, factory *manifests.Factory, config *manifests.Config) *PrometheusAdapterTask {
```
Comment (simonpasquier) [inline pkg/tasks/prometheusadapter.go:22]: (nit)
```suggestion
func NewPrometheusAdapterTask(ctx context.Context, namespace string, client *client.Client, enabled bool, factory *manifests.Factory, config *manifests.Config) *PrometheusAdapterTask {
```

File: pkg/operator/operator.go [hunk 17]
```diff
@@ -749,7 +752,8 @@ func (o *Operator) sync(ctx context.Context, key string) error {
 				newTaskSpec("NodeExporter", tasks.NewNodeExporterTask(o.client, factory)),
 				newTaskSpec("KubeStateMetrics", tasks.NewKubeStateMetricsTask(o.client, factory)),
 				newTaskSpec("OpenshiftStateMetrics", tasks.NewOpenShiftStateMetricsTask(o.client, factory)),
-				newTaskSpec("PrometheusAdapter", tasks.NewPrometheusAdapterTask(ctx, o.namespace, o.client, factory, config)),
+				newTaskSpec("PrometheusAdapter", tasks.NewPrometheusAdapterTask(ctx, o.namespace, o.client, o.prometheusAdapterEnabled, factory, config)),
```
Comment (simonpasquier) [inline pkg/operator/operator.go:755]: (nit) I'd negate o.metricsServerEnabled
```suggestion
				newTaskSpec("PrometheusAdapter", tasks.NewPrometheusAdapterTask(ctx, o.namespace, o.client, !o.metricsServerEnabled, factory, config)),
```

**Comment by jan--f** [comment]
To me this looks good (pending Simons last comments). I'll leave the lgtm to him though.
/label acknowledge-critical-fixes-only

File: pkg/operator/operator.go (hunk 17 above)
Comment (slashpai) [reply pkg/operator/operator.go:755]: I did negate but moved to a new var https://github.com/openshift/cluster-monitoring-operator/pull/2022/files#diff-99c0290799daf9abc6240df64063e20bfaf67b371577b67ac7eec6f4725622ffR439, we don't need that?

Comment (slashpai) [reply pkg/operator/operator.go]: discussed offline, decided to keep only `o.metricsServerEnabled` field

**Review by simonpasquier — COMMENTED** [review]
/lgtm
